packages/pulumi-sdk/src/ApplicationBuilderGeneric.ts

```diff
@@ -0,0 +1,87 @@
+import os from "os";
+import path from "path";
+import fs from "fs";
+import trimNewlines from "trim-newlines";
+import { LocalWorkspace, UpOptions } from "@pulumi/pulumi/automation";
+
+import { PulumiApp } from "./PulumiApp";
+import { getPulumiWorkDir } from "./utils/getPulumiWorkDir";
+import {
+    ApplicationBuilder,
+    ApplicationBuilderConfig,
+    ApplicationStack,
+    ApplicationStackArgs
+} from "./ApplicationBuilder";
+import { ApplicationContext } from "./ApplicationConfig";
+
+export interface ApplicationGenericConfig extends ApplicationBuilderConfig {
+    app(ctx: ApplicationContext): PulumiApp;
+}
+
+export class ApplicationBuilderGeneric extends ApplicationBuilder<ApplicationGenericConfig> {
+    public async createOrSelectStack(args: ApplicationStackArgs): Promise<ApplicationStack> {
+        const PULUMI_SECRETS_PROVIDER = process.env["PULUMI_SECRETS_PROVIDER"];
+
+        // Use ";" when on Windows. For Mac and Linux, use ":".
+        const PATH_SEPARATOR = os.platform() === "win32" ? ";" : ":";
+
+        const relativePath = path.relative(args.projectDir, args.appDir);
+        const pulumiWorkDir = getPulumiWorkDir(args.projectDir, relativePath);
+
+        if (!fs.existsSync(pulumiWorkDir)) {
+            fs.mkdirSync(pulumiWorkDir, { recursive: true });
+        }
+
+        const stack = await LocalWorkspace.createOrSelectStack(
+            {
+                projectName: this.config.name,
+                stackName: args.env,
+                program: async () => {
+                    const app = this.config.app({
+                        env: args.env
+                    });
+
+                    return await app.run();
+                }
+            },
+            {
+                workDir: pulumiWorkDir,
+                projectSettings: {
+                    name: this.config.name,
+                    runtime: "nodejs",
+                    description: this.config.description
+                },
+                secretsProvider: PULUMI_SECRETS_PROVIDER,
```

Comment:
Currently `PULUMI_CONFIG_PASSPHRASE` is taken out of `process.env` and passed to execa `env` options:
![image](https://user-images.githubusercontent.com/2473784/164002571-c1d577ba-b11d-426d-925a-8e9ce7ee1e1b.png)
Because `pulumi` automatically passes all `env` variables down the line, it's unnecessary to do it explicitly, however, it may be nice for the sake of readability.

Reply:
OK 👍🏻